File: 2015/day-01/main.rb
# frozen_string_literal: true

def compare_solutions(expected, actual)
  raise "Expected #{expected} but got #{actual}" unless expected == actual

  puts "Congratulations! Got expected result (#{expected})"
end

def read_file(filename)
  # Get the directory of the currently executing script
  # Join the script directory with the filename
  # Read the content of the file
  File.read(File.join(__dir__, filename))
end

def solve(filename)
  parse_input(filename).sum
end

def solve2(filename)
  floor = 0
  parse_input(filename).each_with_index do |value, id|
    return id + 1 if floor + value == -1

    floor += value
  end
end

def parse_input(filename)
  read_file(filename).each_char.map do |char|
    if char == "("
      1
    else
      -1
    end
  end
end

compare_solutions(-1, solve("test.txt"))
puts "Part1", solve("input.txt")

compare_solutions(5, solve2("test.txt"))
puts "Part2", solve2("input.txt")

File: 2015/day-01/main.py
from pathlib import Path


def compare_solutions(expected, actual):
    if expected != actual:
        raise Exception(f"Expected #{expected} but got #{actual}")

    print(f"Congratulations! Got expected result ({expected})")


def read_file(filename):
    return Path(__file__).parent.joinpath(filename).read_text()


def part1(filename):
    """Solve part 1."""
    return sum(map(char_to_value, list(read_file(filename))))


def part2(filename):
    """Solve part 2."""
    values = map(char_to_value, list(read_file(filename)))
    floor = 0
    for idx, x in enumerate(values):
        if floor + x == -1:
            return idx + 1
        floor += x

    return floor


def char_to_value(char):
    if char == '(':
        return 1
    else:
        return -1


if __name__ == "__main__":
    compare_solutions(-1, part1('test.txt'))
    print(f"Part1, {part1('input.txt')}")

    compare_solutions(5, part2('test.txt'))
    print(f"Part2, {part2('input.txt')}")
